### ISSUE TITLE
The total number of parameters is incorrectly calculated for nested Modules

### ISSUE THREAD

**Issue by andrei-rusu** [TylerYep/torchinfo#122]
**Describe the bug**
The code for computing the total number of parameters is incorrect. It assumes parameters are only contained within leaf nodes, when actually one can have Modules that contain both Parameters and submodules. A good example are graph convolutions in PyTorch Geometric which have both nn.Linear submodules and nn.Parameters for things like biases. The logic marks GCNConvs as non-leaf and hence their direct parameters are never counted.

**To Reproduce**
```python
import torch.nn as nn
import torch_geometric.nn as pyg_nn

class GNN(nn.Module):
    def __init__(self, input_features):
        super().__init__()
        self.g = pyg_nn.GCNConv(input_features, 2)
        self.l = nn.Linear(2,1)
        
    def forward(self, x, edge_index):
        return self.l(self.g(x, edge_index))
    
summary(GNN(x.shape[-1]), input_data=(x, edge_index), depth=1, mode=None)
```

**Expected behavior**
The total number of parameters should include the parameters contained within modules that also have children submodules.

**Comment by TylerYep** [comment]
Thanks for reporting this bug - would appreciate a PR if you get a chance.

Additionally, could you provide a reference output so we know that a potential fix is correct? (i.e. incorrect and then the expected output)

**Comment by andrei-rusu** [comment]
If for the above one sets say ``x = torch.zeros(1000, 13)``, ``edge_index = torch.zeros(2, 50, dtype=torch.long)``, the expected output would be: 'Total params: 31 (as can be calculated by adding up the 'Param #' column, which correctly outputs 28 for the GCNConv and 3 for the final Linear layer). However, ``summary()`` outputs 29 because GCNConv has an internal Linear layer (of 26 weights), but 2 extra learnable Parameters which are not counted (corresponding to bias terms in this case).
This is just a minimal example though, when looking at bigger GNN architectures the total number is way off because every GNN module isn't technically a non-leaf node.

**Comment by andrei-rusu** [comment]
A possible solution could be summing up the total parameters of all layers at depth 1, instead of relying on the ``is_leaf_layer`` check.

**Comment by TylerYep** [comment]
The cleanest solution would be a smarter way of deciding `is_leaf_layer`. Right now we assign:
```
self.is_leaf_layer = not any(self.module.children())
```
and we use this value decide whether to count params or not. I don't think using depth == 1 will always work.

Otherwise, we need some way to distinguish countable parameters from non-leaf modules vs parameters from children modules.

Also, if you could add a simple/minimal reproduction of this bug that uses plain Pytorch, that would help speed up the fix. I really appreciate your investigation on this issue

**Comment by andrei-rusu** [comment]
Yep, I was thinking about that as well but I can't think of something good to replace that boolean that works in all cases. I haven't really completely understood how the metrics calculated are sometimes recursively available at depth 1 but sometimes not. My proposal would work if all the metrics would get set correctly for depth 1 w.r.t. all the children just like the number of parameters is set correctly.

**Comment by andrei-rusu** [comment]
I guess the following would be a good example as it showcases both the error I'm talking about: the Parameter of the 'Inside' module won't be summed up, even though it does appear in the hierarchical view (and this could be fixed in terms of the total parameter numbers by changing the check to `depth == 1``, yet counts like ``macs`` won't always work for other submodule variables like ``ModuleList``) AND an error which most summary libraries out there seem to make: having Parameters defined at depth 0 (this would be a bit trickier to deal with, and ``depth == 1`` won't work)

```python
from torchinfo import summary

class Inside(nn.Module):
    def __init__(self):
        super().__init__()
        self.l_1 = nn.Linear(1,1)
        self.param_1 = nn.Parameter(torch.ones(1))
        
    def forward(self, x):
        return self.l_1(x) * self.param_1

class Model(nn.Module):
    def __init__(self, input_features):
        super().__init__()
        self.l_0 = nn.Linear(2,1)
        self.param_0 = nn.Parameter(torch.ones(2))
        self.inside = Inside()
        
    def forward(self, x):
        return self.inside(self.l_0(x)) * self.param_0
    
summary(Model(x.shape[-1]), input_data=(x,), depth=1)
```

**Comment by TylerYep** [comment]
This has been fixed in https://github.com/TylerYep/torchinfo/commit/2e1f2d1054449be9f0449f61c3625a0f445653c0 and will be released in torchinfo v1.7.0.

Thank you for reporting this issue, and for the detailed investigation! Your example was added to our test cases to ensure the functionality works in the future.
